sign.go

```diff
@@ -29,3 +52,69 @@ func SignInTotoStatement(ctx context.Context, stmt intoto.Statement, provider cl
 
 	return env, nil
 }
+
+func SignInTotoStatementExt(ctx context.Context, stmt intoto.Statement, provider client.OpenIdProvider) (*Envelope, error) {
+	tl := GetTL(ctx)
+
+	// encode in-toto statement
+	payload, err := json.Marshal(stmt)
+	if err != nil {
+		return nil, err
+	}
+	env := new(Envelope)
+	env.Payload = base64.StdEncoding.Strict().EncodeToString(payload)
+	env.PayloadType = intoto.PayloadType
+	encPayload := dsse.PAE(intoto.PayloadType, payload)
+
+	// statement message digest
+	hash := s256(encPayload)
+	hashHex := hex.EncodeToString(hash)
+
+	// generate ephemeral keys to sign message digest
+	signer, err := util.GenKeyPair(jwa.ES256)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %w", err)
+	}
+	sig, err := signer.Sign(rand.Reader, hash, crypto.SHA256)
+	if err != nil {
+		return nil, err
+	}
+	ecPub, ok := signer.Public().(*ecdsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("error casting signer to ecdsa public key")
+	}
+	keyID := s256(append(ecPub.X.Bytes(), ecPub.Y.Bytes()...))
+
+	// generate pk token with message digest and ephemeral signing keys
+	opkClient := client.OpkClient{Op: provider}
+	pkToken, err := opkClient.OidcAuth(ctx, signer, jwa.ES256, map[string]any{"att": hashHex}, true)
+	if err != nil {
+		return nil, fmt.Errorf("error getting PK token: %w", err)
+	}
+	pkTokenJSON, err := json.Marshal(pkToken)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling PK token to JSON: %w", err)
+	}
+
+	// upload to TL
+	entry, err := tl.UploadLogEntry(ctx, pkToken, encPayload, sig, signer)
+	if err != nil {
+		return nil, fmt.Errorf("error uploading TL entry: %w", err)
+	}
+	entryStr := base64.StdEncoding.Strict().EncodeToString(entry)
+
+	// add signature w/ opk extension to dsse envelope
+	env.Signatures = append(env.Signatures, Signature{
+		KeyID: hex.EncodeToString(keyID),                       // ephemeral public key ID
+		Sig:   base64.StdEncoding.Strict().EncodeToString(sig), // ECDSA signature using ephemeral keys
+		Extension: Extension{
+			Kind: OpkSignatureID,
+			Ext: map[string]any{
+				"pkt": pkTokenJSON, // PK token + GQ signature
+				"tl":  entryStr,    // transparency log entry metadata
```

Comment:
Do we need to base64 encode this? Could we just stick it inline like this:
```suggestion
				"tl":  entry,    // transparency log entry metadata
```

Reply:
I originally just assigned it as `[]byte` and it seemed that the JSON marshal handled the conversion automatically.  Here we are using the `StdEncodeding.Strict()` everywhere so it may be best to encode it as that explicitly?

Reply:
Ah sorry, I meant do we need to store this as a base64-encoded string or can it be a JSON object instead?

Reply:
I see now, I'll try that

Reply:
I can but this will be a pretty big change. Are you okay if I open a follow-up PR for storing the entry as JSON? I think this is a good idea.

Reply:
Yes let's do it in a follow-up!